modules/frontend/src/main/scala/smithy4s_codegen/components/CodeEditor.scala

```diff
@@ -65,4 +71,32 @@ class CodeEditor() {
     })
     (icon, errors)
   }
+
+}
+
+/** Writes code to the URL hash and provides a stream of its decoded values.
+  *
+  * Encoding/decoding of code is handled internally.
+  */
+object PermalinkCodec {
+
+  def readOnce(): Option[String] =
+    decode(org.scalajs.dom.window.location.hash)
+
+  val read: EventStream[String] = windowEvents(_.onHashChange)
+    .mapTo(org.scalajs.dom.window.location.hash)
+    .map(decode(_))
+    .collectSome
+
+  def write(code: String): Unit =
+    org.scalajs.dom.window.location.hash = encode(code)
+
```

Comment:
thought: I really wanted something like fs2's `SignallingRef`, basically something that's both settable and can be read "once" or as a stream. `Var` is kind of like that, but can't be created from arbitrary pieces like these or an actual global variable (which `location.hash` is). Happy to take recommendations if you know something like this, or better.